v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,238 @@
+					p.logger.Error("encoding failure: %v, dropping event with decision ID: %v.", err, event.EventV1.DecisionID)
+					continue
+				}
+				if int64(len(event.serialized)) > b.uploadSizeLimitBytes {
+					p.logger.Error("upload event size (%d) exceeds upload_size_limit_bytes (%d), dropping event with decision ID: %v.",
+						int64(len(event.serialized)), b.uploadSizeLimitBytes, event.EventV1.DecisionID)
+					continue
+				}
+
+				p.incrMetric(logNDBDropCounterName)
+				p.logger.Error("ND builtins cache dropped from this event to fit under maximum upload size limits. Increase upload size limit or change usage of non-deterministic builtins.")
+			}
+
+			if !b.payload.capacity(len(event.serialized), b.uploadSizeLimitBytes) {
+				// add the event that doesn't fit back into the buffer
+				b.Push(p, event)
+				break
+			}
+
+			if err := b.payload.add(event); err != nil {
+				// corrupted payload, record failure and reset
+				b.payload.reset()
+				p.incrMetric(logEncodingFailureCounterName)
+				p.logger.Error("encoding failure: %v, dropping event with decision ID: %v.", err, event.EventV1.DecisionID)
+				continue
+			}
+		default:
+			break
+		}
+	}
+
+	if err := b.payload.close(); err != nil {
+		// corrupted payload, record failure and reset
+		b.payload.reset()
+		p.incrMetric(logEncodingFailureCounterName)
+		p.logger.Error("encoding failure: %v, dropping multiple events.")
+		return
+	}
+}
+
+// payload contains the compressed JSON array to be uploaded
+type payload struct {
+	buffer       *bytes.Buffer
+	writer       *gzip.Writer
+	bytesWritten int
+}
+
+func newPayload() *payload {
+	p := payload{}
+	p.reset()
+	return &p
+}
+
+func (p *payload) capacity(eventLen int, uploadSizeLimitBytes int64) bool {
+	return !(int64(eventLen+p.bytesWritten+1) > uploadSizeLimitBytes)
+}
+
+// add writes an event to the JSON array
+func (p *payload) add(event bufferItem) error {
+	switch p.bytesWritten {
+	case 0: // Start new JSON array
+		n, err := p.writer.Write([]byte(`[`))
+		if err != nil {
+			return err
+		}
+		p.bytesWritten += n
+	default: // Append new event to JSON array
+		n, err := p.writer.Write([]byte(`,`))
+		if err != nil {
+			return err
+		}
+		p.bytesWritten += n
+	}
+
+	n, err := p.writer.Write(event.serialized)
+	if err != nil {
+		return err
+	}
+
+	p.bytesWritten += n
+
+	return nil
+}
+
+// close writes the closing bracket to the array
+func (p *payload) close() error {
+	n, err := p.writer.Write([]byte(`]`))
+	if err != nil {
+		return err
+	}
+	p.bytesWritten += n
+
+	if err := p.writer.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// upload closes the JSON array and attempts to send the data to the configured service
+func (p *payload) upload(ctx context.Context, client rest.Client, uploadPath string) (bool, error) {
+	if p.bytesWritten == 0 {
+		return false, nil
+	}
+
+	if err := uploadChunk(ctx, client, uploadPath, p.buffer.Bytes()); err != nil {
+		return false, err
```

Comment:
Looks like the `size` buffer will re-queue events on failed upload. Maybe the `event` buffer should do the same to minimize behavioral differences.

Reply:
PR updated to re-queue events to match existing behavior.